### PR TITLE
feat(ui): Add support for internal entities in network graph

### PR DESCRIPTION
**Pull request by dvail**
## Description

Adds support for internal entities to network graph UI. See testing below for details.

UX Details that need a decision:

- The "internal entities" type uses the same "IP" icon and white square when displayed in the graph. This matches exactly the iconography used for External entities and CIDR blocks. Should this be changed?
- The "Internal to cluster" group in the graph uses the same purple coloring as the "Eternal to cluster" group. Again, should this be changed?
- The internal entities node and internal entities group both display the following in their sidebar:
```
Internal to cluster
Unspecified entity connections within your clusters
```
Does this text read correctly?

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Load the network graph and stackrox namespace. Clicking on `sensor` shows a connection to the general "Internal Entities" node which is contained in its own "Internal to cluster" grouping in the visualization.
image

View the "Flows" tab for `sensor` and see that the `INTERNAL_ENTITIES` type is correctly represented in the sidebar.
image

Click a "Internal entities" entry in the sidebar and ensure the correct node is highlighted in the graph. The sidebar should be updated to reflect connections to and from this entity.
image

Click the "Internal to cluster" group in the graph. See the sidebar containing the list of entities contained in this group. Active traffic counts should match the number of connections displayed within the graph.
image

Clicking the links in the group sidebar should display the correct node in the graph and sidebar:
image

### CODE REVIEW

File: ui/apps/platform/src/Containers/NetworkGraph/cidr/CidrBlockSideBar.tsx
Comment (dvail) [inline ui/apps/platform/src/Containers/NetworkGraph/cidr/CidrBlockSideBar.tsx:1]: The CIDR sidebar, external entity sidebar, and soon-to-be internal entity sidebar all used the exact same structure, minus a couple of text changes. I deleted this component and combined the "ExternalEntitiesSidebar" into a generic component for all 3.

File: ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts
Comment (dvail) [inline ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts:126]: These functions serve the same purpose as the `node.data.type === 'SOME_TYPE'` checks but with the following benefits:

- Using these functions causes the correct node type to be inferred inside the `if` block without the need for casting with `as`
- Keeping a mapping of strings to node types encapsulated here prevents possible errors where `node.data.type` checks and `as NodeType` assertions are used, potentially out of sync
- A 2-arity form of `isNodeOfType(type, node)` and curried `isOfType(type)` form allows us to use these functions easily for checking a single node, or passed to a `filter` for type-safe filtering.

File: ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts
Comment (dvail) [inline ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts:133]: @vikin91 Should this type be the following instead? I wasn't sure if this was a typo or intentional. If the former I can go through and rebase/update this PR after the change on the backend.
```suggestion
    | 'UNKNOWN_INTERNAL_ENTITY'
```

File: ui/apps/platform/src/Containers/NetworkGraph/utils/flowUtils.ts
Comment (dvail) [inline ui/apps/platform/src/Containers/NetworkGraph/utils/flowUtils.ts:104]: If there is a more appropriate field to use as the unique ID, I'll change this.

File: ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts
Comment (vikin91) [reply ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts:133]: The `Unknown Internal Entity` was a predecessor to `Internal Entities`. It should be removed as we abandoned that idea after getting feedback from solution architects.

The idea with `Unknown Internal Entity` was to represent each unknown IP address as a separate node on the graph (named `__ip_10.x.y.z`). That lead pretty quickly to a chaos on the graph as we had too many nodes. Instead, we grouped all of those into single node 'Internal Entities`, which is our plan on how to implement right now.


The code still had the copy of `UNKNOWN_INTERNAL_ENTITY` but I disabled the mechanism in Sensor. 
I should have mentioned that, I am really sorry that I missed that in our meeting :(

Comment (dvail) [reply ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts:133]: No worries - should this type be removed entirely from the PR?

Comment (vikin91) [reply ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts:133]: Yes please, let's remove `UNKNOWN_INTERNAL_ENTITY` and all its typoed versions ;)

Comment (vikin91) [reply ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts]: I will also remove it form Sensor and Central code to keep it clean. Just wondering if I shall do it now and you could rebase, or I would do it after we merge into my feature branch. Do you have any preference?

Comment (dvail) [reply ui/apps/platform/src/Containers/NetworkGraph/types/topology.type.ts]: I think either one should be fine. I already removed it from my branch, so I can rebase again easily if you update yours. I don't imagine there will be any conflicts to worry about.

**Comment by dvail** [comment]
@zhenpesky This is the PR I was talking about, there are a handful of visual additions that can be seen in the screenshots above. Please let me know if these are good as-is, or if there is anything that should be changed to fit better with the current UX.

The new functionality here is that the Network Graph will now display "Internal entities" in the graph for traffic that is internal to the cluster, but the deployment is unknown. This is similar to "External entities", just the opposite. These are the UI changes:

1. For display in the graph, I am using the same "IP" icon and white square used for "External entities" for "Internal entities" as well.
2. The "Internal entities" icon will be contained in its own group in the graph, with a purple label titled "Internal to cluster"
3. When clicking the entity, the sidebar will show a black icon with the text "IE" at the top (I avoided using "I" as it could be confused with "internet")
4. The subtext in the sidebar when viewing these entities is "Unspecified entity connections within your clusters"

**Comment by zhenpesky** [comment]
> @zhenpesky This is the PR I was talking about, there are a handful of visual additions that can be seen in the screenshots above. Please let me know if these are good as-is, or if there is anything that should be changed to fit better with the current UX.
> 
> The new functionality here is that the Network Graph will now display "Internal entities" in the graph for traffic that is internal to the cluster, but the deployment is unknown. This is similar to "External entities", just the opposite. These are the UI changes:
> 
> 1. For display in the graph, I am using the same "IP" icon and white square used for "External entities" for "Internal entities" as well.
> 2. The "Internal entities" icon will be contained in its own group in the graph, with a purple label titled "Internal to cluster"
> 3. When clicking the entity, the sidebar will show a black icon with the text "IE" at the top (I avoided using "I" as it could be confused with "internet")
> 4. The subtext in the sidebar when viewing these entities is "Unspecified entity connections within your clusters"

@dvail  There are a few suggested fixes, let me know what you think: 

1. Have a clearer sentence for 'unspecified'. (example: Unknown pod/deployments that have connections within your cluster)
2. Given that the 'internal entities' are pods/deployments, using the pod icon is more applicable instead of the IP (IP is more related to CIDR). We can simplify the view by having the 'internal entities' shown as a node and not as a group layout.
3. The naming: since we know internal entities mean to be pod/deployments, we can say 'total count + internal pod/deployments' with the 3 stacked box icon.  no borders needed

**Comment by vikin91** [comment]
> @dvail There are a few suggested fixes, let me know what you think:
> 
> 1. Have a clearer sentence for 'unspecified'. (example: Unknown pod/deployments that have connections within your cluster)
> 2. Given that the 'internal entities' are pods/deployments, using the pod icon is more applicable instead of the IP (IP is more related to CIDR). We can simplify the view by having the 'internal entities' shown as a node and not as a group layout.
> 3. The naming: since we know internal entities mean to be pod/deployments, we can say 'total count + internal pod/deployments' with the 3 stacked box icon.  no borders needed

Let me address some of these answers based on our investigations:

1. I agree that the sentence shall be improved. We plan to have a documentation page explaining what are the internal entities with many examples. Important is: they may be deployments and pod, but don't have to be! It may be the k8s API, it may be the Node IP etc. So limiting them to pods is not fully correct.

2. They are not only pods/deployments (although for 80% of cases this is correct). Knowing that, what icon would be best here?

3. Again, let's avoid using words pods and deployments. We shall stick to "entities" in my opinion.

**Review by vikin91 — APPROVED** [review]
For me this state looks great and can be merged into my feat branch. I will then add backend corrections immediately and merge it to master.

We could do the polishing (icon, grouping, texts corrections) as follow-ups. WDYT?

**Comment by dvail** [comment]
@vikin91 that sounds good to me. I created https://issues.redhat.com/browse/ROX-22025 to track the UX tweaks for a follow up. Once we have a UI review of the work here I'll merge into your branch.

**Comment by zhenpesky** [comment]
> For me this state looks great and can be merged into my feat branch. I will then add backend corrections immediately and merge it to master.
> 
> We could do the polishing (icon, grouping, texts corrections) as follow-ups. WDYT?

Thanks for the additional information, I agreed, let's do the polishing later. I will work with David on UX follow-ups.

**Review by sachaudh — APPROVED** [review]
Overall looks good to me. I think I have a little less context than Linda in terms of the progress made on NG since I last worked on it. I'd recommend waiting for her review too.

File: ui/apps/platform/src/Containers/NetworkGraph/TopologyComponent.tsx
Comment (vikin91) [reply ui/apps/platform/src/Containers/NetworkGraph/TopologyComponent.tsx:237]: The future of this is not defined yet. Maybe there will be some in 4.5, but as of Today, I cannot name any example.

**Comment by dvail** [comment]
Thanks @alwayshooin - I'll add your suggestions to the follow up ticket for polishing.

**Pull request merged**
